packages/hardhat-core/test/internal/hardhat-network/provider/modules/hardhat.ts

```diff
@@ -189,6 +189,315 @@ describe("Hardhat module", function () {
+              numberToRpcQuantity(blockNumber),
+              false,
+            ]);
+            assert.isNotNull(block, `expected block ${blockNumber} to exist`);
+            assert.isDefined(block.number);
+            assert.equal(blockNumber, block.number);
+            assert.equal(
+              "0x0000000000000000000000000000000000000000000000000000000000000000",
+              block.parentHash,
+              `expected block's parent hash to be null, not ${block.parentHash}`
+            );
+          };
+
+          const blockCount = 3_000_000_000;
+          const mineBlocks = async () => {
+            await this.ctx.provider.send("hardhat_mine", [
+              numberToRpcQuantity(blockCount),
+            ]);
+          };
+
+          let previousLatestBlockNumber: number;
+
+          beforeEach(async function () {
+            previousLatestBlockNumber = await getLatestBlockNumber();
+            await mineBlocks();
+          });
+
+          it("works at the beginning of the reservation", async function () {
+            await getAndAssertBlock(previousLatestBlockNumber + 1);
+          });
+
+          it("works in the middle of the reservation", async function () {
+            await getAndAssertBlock(
+              previousLatestBlockNumber + 1 + Math.floor(blockCount / 2)
+            );
+          });
+
+          it("works at the end of the reservation", async function () {
+            await getAndAssertBlock(previousLatestBlockNumber + blockCount);
+          });
+
+          it("works several times over within the reservation", async function () {
+            for (
+              let blockNumber = previousLatestBlockNumber + blockCount;
+              blockNumber > previousLatestBlockNumber;
+              blockNumber = Math.floor(blockNumber / 2)
+            ) {
+              await getAndAssertBlock(blockNumber);
+            }
+          });
+        });
+
+        describe("should increment the block number", async () => {
+          it("when not given any arguments", async () => {
+            const latestBlockNumber = await getLatestBlockNumber();
+            await this.ctx.provider.send("hardhat_mine");
+            assert.equal(await getLatestBlockNumber(), latestBlockNumber + 1);
+          });
+          it("when mining 1000 blocks", async () => {
+            const latestBlockNumber = await getLatestBlockNumber();
+            await this.ctx.provider.send("hardhat_mine", [
+              numberToRpcQuantity(1000),
+            ]);
+            assert.equal(
+              await getLatestBlockNumber(),
+              latestBlockNumber + 1000
+            );
+          });
+        });
+
+        describe("should reflect timestamps properly", function () {
+          it("with only one hardhat_mine invocation", async () => {
+            const originalLatestBlockNumber = await getLatestBlockNumber();
+            const timestampBefore = rpcQuantityToNumber(
```

Comment:
Maybe it's worth creating a `getBlockTimestamp` helper to make this test more readable.

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/2032/commits/b557db173f5a4ef56205fbfa7a5562ecf52222fe